src/aws-cpp-sdk-core/include/smithy/interceptor/impl/ChecksumInterceptor.h

```diff
@@ -0,0 +1,209 @@
+                const auto& modeledRequest = context.GetModeledRequest();
+                if (httpRequest == nullptr)
+                {
+                    return Aws::Client::AWSError<Aws::Client::CoreErrors>{Aws::Client::CoreErrors::VALIDATION,
+                        "ValidationErrorException",
+                        "Checksum request validation missing request",
+                        false};
+                }
+                const auto serviceSpecificParams = modeledRequest.GetServiceSpecificParameters();
+                // Request checksums
+                Aws::String checksumAlgorithmName = Aws::Utils::StringUtils::ToLower(modeledRequest.GetChecksumAlgorithmName().c_str());
+                if (serviceSpecificParams) {
+                    auto requestChecksumOverride = serviceSpecificParams->parameterMap.find("overrideChecksum");
+                    if (requestChecksumOverride != serviceSpecificParams->parameterMap.end()) {
+                        checksumAlgorithmName = requestChecksumOverride->second;
+                    }
+                }
+
+                bool shouldSkipChecksum = serviceSpecificParams &&
+                                          serviceSpecificParams->parameterMap.find("overrideChecksumDisable") !=
+                                          serviceSpecificParams->parameterMap.end();
+
+                //Check if user has provided the checksum algorithm
+                if (!checksumAlgorithmName.empty() && !shouldSkipChecksum)
+                {
+                    // Check if user has provided a checksum value for the specified algorithm
+                    const Aws::String checksumType = "x-amz-checksum-" + checksumAlgorithmName;
+                    const Aws::Http::HeaderValueCollection &headers = httpRequest->GetHeaders();
+                    const auto checksumHeader = headers.find(checksumType);
+                    bool checksumValueAndAlgorithmProvided = checksumHeader != headers.end();
+
+                    // For non-streaming payload, the resolved checksum location is always header.
+                    // For streaming payload, the resolved checksum location depends on whether it is an unsigned payload, we let AwsAuthSigner decide it.
+                    if (modeledRequest.IsStreaming() && checksumValueAndAlgorithmProvided)
+                    {
+                        const auto hash = Aws::MakeShared<Aws::Utils::Crypto::PrecalculatedHash>(CHECKSUM_INTERCEPTOR_LOG_TAG, checksumHeader->second);
+                        httpRequest->SetRequestHash(checksumAlgorithmName,hash);
+                    }
+                    else if (checksumValueAndAlgorithmProvided){
+                        httpRequest->SetHeaderValue(checksumType, checksumHeader->second);
+                    }
+                    else if (checksumAlgorithmName == "crc32")
+                    {
+                        if (modeledRequest.IsStreaming())
```

Comment:
cleanup: this block can be wrapped in a lambda accepting std:function which can be Aws::Utils::HashingUtils::CalculateCRC32 or Aws::Utils::HashingUtils::CalculateCRC32C or Aws::Utils::HashingUtils::CalculateSHA256
Rest is the same. Just easy to read and more code reuse. 

Also we should move away from hardcoded strings like "crc32" unless they are constants defined or just use enum.
And move to switch case, which is much faster than if/else

Reply:
from my other comment

> all of the code in this file is copied from [the existing implentation](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/client/AWSClient.cpp#L799). I will refactor the logic in the next PR. This PR more or less jsut refactors the code out to a single location before handling the actual logical refactor. So while we can address comments on the logic here, a lot will be changing to it in the next PR.

This code could be cleaned up in a lot of different ways, and we plan on it. Its actually the next PR im going to raise. I dont like how this is strucutred and we ill be moving away from this giant if-else block alltogether. This PR is more focused on moving it out of the client and compartmentalizing it, then cleaning it up